qiime2/sdk/action.py

```diff
@@ -276,6 +276,45 @@ def _build_annotations(self):
 
         return annotations
 
+    def _build_numpydoc(self):
+        numpydoc = []
+        numpydoc.append(textwrap.fill(self.name, width=79))
+        numpydoc.append(textwrap.fill(self.description, width=79))
+
+        sig = self.signature
+        params = collections.OrderedDict()
+        params.update(sig.inputs)
+        params.update(sig.parameters)
+
+        parameters = self._build_section("Parameters", params)
+        returns = self._build_section("Returns", sig.outputs)
+
+        for section in (parameters, returns):
+            if section:
+                numpydoc.append(section)
+
+        return '\n\n'.join(numpydoc) + '\n'
+
+    def _build_section(self, header, iterable):
+        section = []
+
+        if iterable:
+            section.append(header)
+            section.append('-'*len(header))
+            for key, value in iterable.items():
+                section.append(
+                    "{item} : {type}".format(item=key, type=value.qiime_type))
+                if value.has_default():
+                    section[-1] += ", optional"
```

Comment:
instead of reindexing the list, how about you append the finished string.

```python
figure_out_a_name = "{item} : {type}".format(item=key, type=value.qiime_type))
if value.has_default():
    figure_out_a_name += ", optional"
section.append(figure_out_a_name)
```